File: csharp/src/SpringNBTLibrary/World/Cuboid.cs
using System.Globalization;

namespace SpringNBTLibrary.World;

/// <summary>
/// ブロック座標の直方体な範囲
/// </summary>
/// <remarks>
/// <para>両端を含む</para>
/// <para><c>Of(0, 0, 0, 0, 0, 0)</c> は 1 ブロック</para>
/// <para>範囲内のブロックを順に処理したいときに使う</para>
/// </remarks>
public sealed class Cuboid : IEquatable<Cuboid>
{
    /// <summary>両端の座標から作る</summary>
    /// <remarks>大小の順序は問わない
    /// 内部で小さいほうを最小に揃える</remarks>
    public Cuboid(BlockPos first, BlockPos second)
    {
        ArgumentNullException.ThrowIfNull(first);
        ArgumentNullException.ThrowIfNull(second);

        MinX = Math.Min(first.X, second.X);
        MinY = Math.Min(first.Y, second.Y);
        MinZ = Math.Min(first.Z, second.Z);
        MaxX = Math.Max(first.X, second.X);
        MaxY = Math.Max(first.Y, second.Y);
        MaxZ = Math.Max(first.Z, second.Z);
    }

    /// <summary>両端の座標から作る</summary>
    public static Cuboid Of(int x1, int y1, int z1, int x2, int y2, int z2) =>
        new Cuboid(new BlockPos(x1, y1, z1), new BlockPos(x2, y2, z2));

    /// <summary>X の最小値</summary>
    public int MinX { get; }

    /// <summary>Y の最小値</summary>
    public int MinY { get; }

    /// <summary>Z の最小値</summary>
    public int MinZ { get; }

    /// <summary>X の最大値（含む）</summary>
    public int MaxX { get; }

    /// <summary>Y の最大値（含む）</summary>
    public int MaxY { get; }

    /// <summary>Z の最大値（含む）</summary>
    public int MaxZ { get; }

    /// <summary>X 方向の長さ</summary>
    public int SizeX => MaxX - MinX + 1;

    /// <summary>Y 方向の長さ</summary>
    public int SizeY => MaxY - MinY + 1;

    /// <summary>Z 方向の長さ</summary>
    public int SizeZ => MaxZ - MinZ + 1;

    /// <summary>含まれるブロックの個数</summary>
    public long Volume => (long)SizeX * SizeY * SizeZ;

    /// <summary>その座標が範囲に含まれるか</summary>
    public bool Contains(int x, int y, int z) =>
        x >= MinX && x <= MaxX && y >= MinY && y <= MaxY && z >= MinZ && z <= MaxZ;

    /// <summary>
    /// 範囲内の座標を順に返す
    /// </summary>
    /// <remarks>並びは Y、Z、X の順で、X がいちばん内側で動く</remarks>
    public IEnumerable<BlockPos> Positions()
    {
        // 内側から X が動くので、同じチャンクの並びを続けて触れる
        for (int y = MinY; y <= MaxY; y++)
        {
            // Z を進めながら、その行の X を端から端まで返す
            for (int z = MinZ; z <= MaxZ; z++)
            {
                // X はチャンク内で連続するので、まとめて触れる
                for (int x = MinX; x <= MaxX; x++)
                {
                    yield return new BlockPos(x, y, z);
                }
            }
        }
    }

    /// <inheritdoc/>
    public bool Equals(Cuboid? other)
    {
        if (other is null)
        {
            return false;
        }

        return other.MinX == MinX && other.MinY == MinY && other.MinZ == MinZ
            && other.MaxX == MaxX && other.MaxY == MaxY && other.MaxZ == MaxZ;
    }

    /// <inheritdoc/>
    public override bool Equals(object? obj) => Equals(obj as Cuboid);

    /// <inheritdoc/>
    public override int GetHashCode() => HashCode.Combine(MinX, MinY, MinZ, MaxX, MaxY, MaxZ);

    /// <inheritdoc/>
    public override string ToString() => string.Create(
        CultureInfo.InvariantCulture, $"({MinX}, {MinY}, {MinZ})-({MaxX}, {MaxY}, {MaxZ})");
}

File: java/src/main/java/io/github/scriptarts/springnbt/world/Cuboid.java
package io.github.scriptarts.springnbt.world;

import java.util.Iterator;
import java.util.NoSuchElementException;
import java.util.Objects;

/**
 * ブロック座標の直方体な範囲
 *
 * <p>両端を含む。{@code of(0, 0, 0, 0, 0, 0)} は 1 ブロック
 *
 * <p>範囲内のブロックを順に処理したいときに使う
 */
public final class Cuboid {

    private final int minX;
    private final int minY;
    private final int minZ;
    private final int maxX;
    private final int maxY;
    private final int maxZ;

    /**
     * 両端の座標から作る
     *
     * <p>大小の順序は問わない。内部で小さいほうを最小に揃える
     *
     * @param first 端の座標
     * @param second もう一方の端の座標
     */
    public Cuboid(BlockPos first, BlockPos second) {
        Objects.requireNonNull(first, "first");
        Objects.requireNonNull(second, "second");

        minX = Math.min(first.x(), second.x());
        minY = Math.min(first.y(), second.y());
        minZ = Math.min(first.z(), second.z());
        maxX = Math.max(first.x(), second.x());
        maxY = Math.max(first.y(), second.y());
        maxZ = Math.max(first.z(), second.z());
    }

    /**
     * 両端の座標から作る
     *
     * @param x1 一方の端のX座標
     * @param y1 一方の端のY座標
     * @param z1 一方の端のZ座標
     * @param x2 もう一方の端のX座標
     * @param y2 もう一方の端のY座標
     * @param z2 もう一方の端のZ座標
     * @return 範囲
     */
    public static Cuboid of(int x1, int y1, int z1, int x2, int y2, int z2) {
        return new Cuboid(new BlockPos(x1, y1, z1), new BlockPos(x2, y2, z2));
    }

    /**
     * X の最小値
     *
     * @return 座標
     */
    public int minX() {
        return minX;
    }

    /**
     * Y の最小値
     *
     * @return 座標
     */
    public int minY() {
        return minY;
    }

    /**
     * Z の最小値
     *
     * @return 座標
     */
    public int minZ() {
        return minZ;
    }

    /**
     * X の最大値（含む）
     *
     * @return 座標
     */
    public int maxX() {
        return maxX;
    }

    /**
     * Y の最大値（含む）
     *
     * @return 座標
     */
    public int maxY() {
        return maxY;
    }

    /**
     * Z の最大値（含む）
     *
     * @return 座標
     */
    public int maxZ() {
        return maxZ;
    }

    /**
     * X 方向の長さ
     *
     * @return 長さ
     */
    public int sizeX() {
        return maxX - minX + 1;
    }

    /**
     * Y 方向の長さ
     *
     * @return 長さ
     */
    public int sizeY() {
        return maxY - minY + 1;
    }

    /**
     * Z 方向の長さ
     *
     * @return 長さ
     */
    public int sizeZ() {
        return maxZ - minZ + 1;
    }

    /**
     * 含まれるブロックの個数
     *
     * @return 個数
     */
    public long volume() {
        return (long) sizeX() * sizeY() * sizeZ();
    }

    /**
     * その座標が範囲に含まれるか
     *
     * @param x X座標
     * @param y Y座標
     * @param z Z座標
     * @return 含まれれば true
     */
    public boolean contains(int x, int y, int z) {
        return x >= minX && x <= maxX && y >= minY && y <= maxY && z >= minZ && z <= maxZ;
    }

    /**
     * 範囲内の座標を順に返す
     *
     * <p>並びは Y、Z、X の順で、X がいちばん内側で動く
     *
     * @return 座標の並び
     */
    public Iterable<BlockPos> positions() {
        return () -> new Iterator<BlockPos>() {
            private int x = minX;
            private int y = minY;
            private int z = minZ;
            private boolean done = volume() == 0;

            @Override
            public boolean hasNext() {
                return !done;
            }

            @Override
            public BlockPos next() {
                if (done) {
                    throw new NoSuchElementException();
                }

                BlockPos current = new BlockPos(x, y, z);
                x++;

                // 内側から X が動くので、同じチャンクの並びを続けて触れる
                if (x > maxX) {
                    x = minX;
                    z++;

                    if (z > maxZ) {
                        z = minZ;
                        y++;

                        if (y > maxY) {
                            done = true;
                        }
                    }
                }

                return current;
            }
        };
    }

    @Override
    public boolean equals(Object other) {
        if (!(other instanceof Cuboid box)) {
            return false;
        }

        return box.minX == minX && box.minY == minY && box.minZ == minZ
                && box.maxX == maxX && box.maxY == maxY && box.maxZ == maxZ;
    }

    @Override
    public int hashCode() {
        return Objects.hash(minX, minY, minZ, maxX, maxY, maxZ);
    }

    @Override
    public String toString() {
        return "(" + minX + ", " + minY + ", " + minZ + ")-("
                + maxX + ", " + maxY + ", " + maxZ + ")";
    }
}
